File: 3612-process-string-with-special-operations-i/3612-process-string-with-special-operations-i.go
func processStr(s string) string {
	result := []byte{}
	for i := 0; i < len(s); i++ {
		if s[i] == '*' {
			if len(result) > 0 {
				result = result[:len(result)-1]
			}
		} else if s[i] == '#' {
            result = append(result, result...)
        } else if s[i] == '%' {
            result = reverse(result)
        } else {
            result = append(result, s[i])
        }
	}

    return string(result)
}

func reverse(s []byte) []byte {
	temp := make([]byte, len(s))
	for i := 0; i < len(s); i++ {
		temp[i] = s[len(s) - i - 1]
	}
	return temp
}

// string
// time: O(nk)
// space: O(n)

File: 3612-process-string-with-special-operations-i/3612-process-string-with-special-operations-i.py
class Solution:
    def processStr(self, s: str) -> str:
        result = []
        for c in s:
            if c == "*":
                if result:
                    result.pop()
            elif c == "#":
                result += result
            elif c == "%":
                result.reverse()
            else:
                result.append(c)

        return "".join(result)

# string
# time: O(nk)
# space: O(n)